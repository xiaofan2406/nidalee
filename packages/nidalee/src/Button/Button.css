[data-ndl-button] {
  display: inline-flex;
  justify-content: center;
  align-items: center;

  font-size: 1rem;
  color: var(--ndl-white);
  border-radius: 1px;

  &:focus-visible {
    outline: var(--ndl-outline);
  }

  &:disabled {
    opacity: 0.6;
    pointer-events: none;
  }
}

[data-ndl-button-variant='text'] {
  &[data-ndl-accented] {
    color: var(--ndl-accent-4);
  }
}

[data-ndl-button-variant='button'] {
  padding: 0.5rem 1rem;
  background-color: var(--ndl-black-7);

  &:hover {
    background-color: var(--ndl-black-5);
  }

  &:active {
    background-color: var(--ndl-black-6);
  }

  &[data-ndl-accented] {
    background-color: var(--ndl-accent-8);

    &:hover {
      background-color: var(--ndl-accent-5);
    }

    &:active {
      background-color: var(--ndl-accent-6);
    }
  }
}
