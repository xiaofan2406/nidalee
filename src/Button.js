/* @flow */
import * as React from 'react';
import styled from 'react-emotion';
import Color from 'color';
import { theme, fontSizes } from './styles';
import { I } from './Icon';
import Spinner from './Spinner';

const EButton = styled.button`
  display: inline-flex;
  align-items: center;
  vertical-align: top;

  outline: none;
  cursor: pointer;
  user-select: none;
  &:focus {
    background-color: ${({ color }) =>
      Color(color)
        .darken(0.2)
        .string()};
  }

  font-family: ${theme.fontFamily};
  font-size: ${({ size }) => size}px;
  line-height: 1.2;

  height: ${({ height }) => height}px;
  padding: 0.6em 0.8em;
  margin: 0px 8px 0px 0px;

  border: 1px solid transparent;
  border-radius: 2px;

  & > *,
  & > ${I} {
    margin-left: 0px;
    margin-right: 6px;
    &:last-child {
      margin-right: 0px;
    }
    &:first-child {
      margin-left: 0px;
    }
  }

  color: ${theme.textColor};
  background-color: ${({ color }) =>
    color !== theme.eleBgColor ? color : 'transparent'};
  box-shadow: ${({ color }) =>
    color !== theme.eleBgColor
      ? '0px 1px 3px 0px rgba(0, 0, 0, 0.4), 0px 2px 2px 0px rgba(0, 0, 0, 0.2), 0px 2px 0px 0px rgba(0, 0, 0, 0.1)'
      : 'unset'};

  background-position: center;
  transition: background 0.6s;
  &:hover {
    background: ${({ color }) => color}
      radial-gradient(
        circle,
        transparent 1%,
        ${({ color }) =>
            Color(color)
              .lighten(0.2)
              .string()}
          1%
      )
      center/15000%;
  }
  &:active {
    background-color: ${({ color }) =>
      Color(color)
        .darken(0.2)
        .string()};
    background-size: 100%;
    transition: background 0s;
  }
`;

class Button extends React.Component<ButtonProps> {
  static defaultProps: ButtonDefaultProps = {
    primary: false,
    size: 'regular',
    color: theme.eleBgColor,
    showSpinner: false,
  };

  get size(): number {
    return fontSizes[this.props.size];
  }

  get height(): number {
    return ({ small: 30, regular: 36, large: 46 }: {
      [key: Size]: number,
    })[this.props.size];
  }

  get color(): string {
    const { primary, color } = this.props;
    return primary ? theme.primaryColor : color;
  }

  renderSpinner = () => {
    const { showSpinner } = this.props;

    return showSpinner ? (
      <Spinner
        size={this.size + 2}
        color={theme.textColor}
        className="spinner"
      />
    ) : null;
  };

  render() {
    const { children, primary, color, showSpinner, ...rest } = this.props;

    return (
      <EButton
        {...rest}
        color={this.color}
        size={this.size}
        height={this.height}
      >
        {this.renderSpinner()}
        {React.Children.map(
          children,
          child => (typeof child === 'string' ? <span>{child}</span> : child)
        )}
      </EButton>
    );
  }
}

export default Button;
