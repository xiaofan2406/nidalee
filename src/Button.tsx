import * as React from 'react';

interface ButtonProps
  extends React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  > {}

function Button(props: ButtonProps) {
  return <button {...props} />;
}

export default Button;
