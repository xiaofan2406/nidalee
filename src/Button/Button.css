.ndl-button {
  display: inline-flex;
  justify-content: center;
  align-items: center;

  font-size: 1rem;
  color: var(--ndl-white);
  border-radius: 1px;

  &.mode-button {
    padding: 0.5rem 1rem;
    background-color: var(--ndl-black-7);

    &:hover {
      background-color: var(--ndl-black-5);
    }

    &:active {
      background-color: var(--ndl-black-6);
    }

    &.accented {
      background-color: var(--ndl-accent-8);

      &:hover {
        background-color: var(--ndl-accent-5);
      }

      &:active {
        background-color: var(--ndl-accent-6);
      }
    }
  }

  &.mode-text {
    &.accented {
      color: var(--ndl-accent-4);
    }
  }

  &:focus-visible {
    outline: var(--ndl-outline);
  }

  &:disabled {
    opacity: 0.6;
    pointer-events: none;
  }
}
