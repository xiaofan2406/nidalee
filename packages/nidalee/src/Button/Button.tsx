import {cx} from '../utils';

import './Button.css';

export type ButtonMode = 'button' | 'text';

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  accented?: boolean;
  mode?: ButtonMode;
}

// https://www.w3.org/TR/wai-aria-practices-1.1/#button
export const Button = ({
  accented,
  mode = 'button',
  className,
  ...rest
}: ButtonProps) => {
  return (
    <button
      type="button"
      {...rest}
      className={cx(
        'ndl-button',
        `mode-${mode}`,
        accented && 'accented',
        className
      )}
    />
  );
};
