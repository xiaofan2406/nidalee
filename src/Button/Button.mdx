---
name: Button
---

import { Playground, PropsTable } from 'docz';
import Button from './Button';
import Spinner from '../Spinner/Spinner';

# Button

## Props

<PropsTable of={Button} />

In addition, it supports all HTML `button` element props.

## Basic usage

<Playground>
  <Button>Submit</Button>
  <Button size="small">Submit</Button>
  <Button size="large">Submit</Button>
  <Button color="red">Submit</Button>
</Playground>
