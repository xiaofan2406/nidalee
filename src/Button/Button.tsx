import * as React from 'react';

import './Button.css';

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  accented?: boolean;
}

// https://www.w3.org/TR/wai-aria-practices-1.1/#button
export const Button = ({accented, ...rest}: ButtonProps) => {
  return (
    <button
      type="button"
      {...rest}
      data-ndl-button=""
      data-accented={accented ? '' : undefined}
    />
  );
};
