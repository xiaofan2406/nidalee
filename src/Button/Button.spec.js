import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import Button from './Button';

it('takes html button props', () => {
  const onClick = jest.fn();
  const { getByText } = render(<Button onClick={onClick}>Click</Button>);

  fireEvent.click(getByText('Click'));
  expect(onClick).toHaveBeenCalledTimes(1);
});

it('sets background color', () => {
  const { getByText } = render(<Button color="red">Click</Button>);

  expect(getByText('Click')).toHaveStyle('background-color: red');
});
