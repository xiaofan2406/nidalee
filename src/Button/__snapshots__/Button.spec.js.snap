// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component matches snapshot 1`] = `
<button
  class="css-sabna4-Button"
>
  <span>
    Click
  </span>
</button>
`;
